lib/plug/csrf_protection.ex

```diff
@@ -103,17 +109,27 @@ defmodule Plug.CSRFProtection do
 
   def init(opts), do: opts
 
-  def call(conn, _opts) do
+  def call(conn, opts) do
     csrf_token = get_session(conn, "_csrf_token")
     Process.put(:plug_csrf_token, csrf_token)
-
+    
     if not verified_request?(conn, csrf_token) do
-      raise InvalidCSRFTokenError
+      conn = case Keyword.get(opts, :with, :exception) do
+        :exception   -> raise InvalidCSRFTokenError
+        :nil_session -> clear_session_for_conn(conn)
+        _ -> raise ArgumentError, message: "Option :with should be one of :exception or :nil_session"
+      end
     end
 
     register_before_send(conn, &ensure_same_origin_and_csrf_token!(&1, csrf_token))
   end
 
+  #do not persist changes to this session
+  defp clear_session_for_conn(conn) do
+    conn = configure_session(conn, ignore: true)
```

Comment:
Should we ignore or drop the session? If we ignore, doesn't it mean the next request will still have a session with the invalid csrf token in it?